### DISCUSSION TITLE
option --pythonVersion not recognized

### DISCUSSION THREAD

**Discussion by ernstpijper** [DIRACGrid/DIRAC#7673] [Q&A] [answered]
DIRAC py3 server 7.3.17, py3 pilot jobs and py3 diracos2 client

Still running an old version on our dirac production server but I'm in the process of updating my dirac test machine to the latest version :)

Since a couple of days pilot job crash with the following error message:

option --pythonVersion not recognized

The command that is being run is:

Executing: python dirac-pilot.py -S DIRAC-Production --pythonVersion=3 -ddd -N arc01.gina.surfsara.nl -Q long-4core -n GRID.SURF.nl --modules=https://github.com/chaen/DIRAC.git:::DIRAC:::rel-v7r3_FEAT_proxyStrength

This started all of a sudden. I asked the admins of the compute cluster if something changed that could explain this. But the dirac command that is being run is not the one on the worker node but the one in the python environment setup by the pilot, right?

Anyway, hope you can tell me where to look.

Kind regards,
Ernst Pijper

**Comment by aldbr** [comment]
Hi,

We [recently](https://github.com/DIRACGrid/DIRAC/discussions/7512#discussioncomment-9725850) merged the following PR into the master branch of the Pilot: https://github.com/DIRACGrid/Pilot/pull/235
It basically drops the possibility of having a python2 installation of Dirac within a pilot.

And I think we don't maintain `v7r3` anymore.
The best you can do is likely to set the SiteDirector `Python3Pilots` option to `False` in the CS. 

https://github.com/DIRACGrid/DIRAC/blob/5938d103018eeb0f4a46b3418d7400e7d2f68169/src/DIRAC/WorkloadManagementSystem/Agent/SiteDirector.py#L996-L997

It should prevent having `--pythonVersion=3` in the options of the pilot.
Since you are already using python3 pilot jobs, everything should be fine.

Cheers

**Reply by ernstpijper** [reply]
Hi aldbr,

Thanks for helping me out. We suffered from network problems last week so I couldn't test your solution. 
So I set

Python3Pilots = False

in the DIRAC config and ran a simple job. I can see it works but it fails somewhere else now and it's a mystery to me why:

This is from the pilot output:
```
2024-06-17T10:43:34.809303Z DEBUG [InstallDIRAC] Return code of bash -c 'source $DIRAC_RC_PATH && env': 0
2024-06-17T10:43:34.810383Z INFO [InstallDIRAC] Executing command pip install  git+https://github.com/chaen/DIRAC.git@rel-v7r3_FEAT_proxyStrength#egg=DIRAC[pilot]
Collecting DIRAC (from DIRAC[pilot])
  Cloning https://github.com/chaen/DIRAC.git (to revision rel-v7r3_FEAT_proxyStrength) to /tmp/pip-install-xktboq6c/dirac_075f05b1dbcd400ab730692bae074b2f

2024-06-17T10:43:39.677368Z DEBUG [InstallDIRAC] Return code of pip install  git+https://github.com/chaen/DIRAC.git@rel-v7r3_FEAT_proxyStrength#egg=DIRAC[pilot]: 1
2024-06-17T10:43:39.677509Z ERROR [InstallDIRAC] Could not pip install  git+https://github.com/chaen/DIRAC.git@rel-v7r3_FEAT_proxyStrength#egg=DIRAC[pilot] [ERROR 1]    
2024-06-17T10:43:39.677549Z INFO [InstallDIRAC] Content of pilot.cfg
```

and this from the pilot error:

```
https://nl-dirac-web.grid.surfsara.nl unreacheable (this is normal!)

  Running command git clone --filter=blob:none --quiet https://github.com/chaen/DIRAC.git /tmp/pip-install-xktboq6c/dirac_075f05b1dbcd400ab730692bae074b2f
  WARNING: Did not find branch or tag 'rel-v7r3_FEAT_proxyStrength', assuming revision or ref.
  Running command git checkout -q rel-v7r3_FEAT_proxyStrength
  error: pathspec 'rel-v7r3_FEAT_proxyStrength' did not match any file(s) known to git
  error: subprocess-exited-with-error
  
   git checkout -q rel-v7r3_FEAT_proxyStrength did not run successfully.
   exit code: 1
  > See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

 git checkout -q rel-v7r3_FEAT_proxyStrength did not run successfully.
 exit code: 1
> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.

Traceback (most recent call last):
  File "dirac-pilot.py", line 115, in <module>
    command.execute()
  File "/tmp/k8WKDmO0Od5niyfV3nXQVtOmABFKDmABFKDmtUGKDmABFKDmFEyalm/DIRAC_0WBlZVpilot/pilotCommands.py", line 81, in wrapper
    return func(self)
  File "/tmp/k8WKDmO0Od5niyfV3nXQVtOmABFKDmABFKDmtUGKDmABFKDmFEyalm/DIRAC_0WBlZVpilot/pilotCommands.py", line 427, in execute
    self._localInstallDIRAC()
  File "/tmp/k8WKDmO0Od5niyfV3nXQVtOmABFKDmABFKDmtUGKDmABFKDmFEyalm/DIRAC_0WBlZVpilot/pilotCommands.py", line 382, in _localInstallDIRAC
    self.exitWithError(retCode)
  File "/tmp/k8WKDmO0Od5niyfV3nXQVtOmABFKDmABFKDmtUGKDmABFKDmFEyalm/DIRAC_0WBlZVpilot/pilotTools.py", line 797, in exitWithError
    with open("pilot.cfg") as f:
IOError: [Errno 2] No such file or directory: 'pilot.cfg'
```

In my DIRAC config I have the following:


```
          arc01.gina.surfsara.nl
          {
            Modules = https://github.com/chaen/DIRAC.git:::DIRAC:::rel-v7r3_FEAT_proxyStrength
            CEType = ARC
            OS = CentOS_Core_7.9
            architecture = x86_64
            Queues
            {
```

Feature rel-v7r3_FEAT_proxyStrength solves a problem with the proxy strength.

Hope you can help me out once more.

Regards,
Ernst

**Reply by andresailer** [reply]
That branch was removed last week (https://github.com/DIRACGrid/DIRAC/pull/6043#event-13115597292)
![image](https://github.com/DIRACGrid/DIRAC/assets/2115797/dc14c5a2-58bd-413f-abc5-c2b7e3acb13e)

@chaen Can you restore the branch (and protect it) in your fork? (click on the link above and it might still let you restore it?

**Reply by andresailer** [reply]
There is a copy of the branch now https://github.com/andresailer/DIRAC/tree/feat_PROXY_PR
Same SHA as https://github.com/DIRACGrid/DIRAC/pull/6043#event-6582925415
```
git fetch origin pull/6043/head:feat_PROXY_PR
git switch feat_PROXY_PR
git push downstream feat_PROXY_PR
```
(origin is DIRACGrid, downstream is my repo)
@ernstpijper  please push this branch into a fork that you control

**Reply by ernstpijper** [reply]
Thanks Andre,

I'll do my best to make it work with limited knowledge of git. If I run into problems I hope I can bother you some more :)

**Reply by ernstpijper** [reply]
Hi Andre,

Im running into the following problem and I don't know how to solve it:

```
git push origin feat_PROXY_PR
Enumerating objects: 213772, done.
Counting objects: 100% (213772/213772), done.
Delta compression using up to 8 threads
Compressing objects: 100% (49870/49870), done.
remote: error: object 82a157e00defa3cd032dd62e4a0c659265a7c868: badEmail: invalid author/committer line - bad email
remote: fatal: fsck error in packed object
error: remote unpack failed: index-pack failed
To github.com:ernstpijper/DIRAC.git
 ! [remote rejected]     feat_PROXY_PR -> feat_PROXY_PR (failed)
error: failed to push some refs to 'github.com:ernstpijper/DIRAC.git'
```

These are the contents of that object:

```
git cat-file -p 82a157e00defa3cd032dd62e4a0c659265a7c868
tree 4d7fd477d6b97c0de97327ce012e29eb11633137
parent 0eb1b7affe0e23a154b66ae37d23194dbbe85ca5
author Adrià Casajús <<adria@ecm.ub.es> 1304668768 +0200
committer Adrià Casajús <<adria@ecm.ub.es> 1304668768 +0200

CHANGE: Defaults section is to be now LocalInstallation	
```

Not sure what the problem is with the email address/committer name.

**Reply by ernstpijper** [reply]
Hi Andre,

For now I'm using the feat_PROXY_PR branch in your account. That's works just fine. Until I can push this branch into my account, I hope you will not delete it.

**Reply by andresailer** [reply]
I hope you have ssh configured for github. then this should be it
```
mkdir -p /tmp/foo/bar
cd /tmp/foo/bar
git clone git@github.com:ernstpijper/DIRAC.git
cd DIRAC
git remote add andre https://github.com/andresailer/DIRAC.git
git fetch andre
git checkout -t andre/feat_PROXY_PR 
git push origin feat_PROXY_PR 
````

**Reply by andresailer** [reply]
Ahh, sorry, I missed your previous report about the issue.

**Reply by andresailer** [reply, marked as answer]
Please delete your https://github.com/ernstpijper/DIRAC repository 

And then "fork" the existing DIRAC

https://github.com/DIRACGrid/DIRAC/fork

instead.
And then use the commands I posted above, for example.

**Reply by ernstpijper** [reply]
Hi Andre,

Everyhting is working again. Thank you so much for your help

Ernst

**Reply by andresailer** [reply]
Great! Thanks for confirming!

**Comment by IgorPelevanyuk** [comment]
Hi, 
Thanks @aldbr for the direction for the further search.

I have tried that solution. But for version 8.0.36 it did not work. The problem is that in that version the option is appended automatically. 

https://github.com/DIRACGrid/DIRAC/blob/fa1dceeb80b13d2123089eb0fdaa0abb7577b92b/src/DIRAC/WorkloadManagementSystem/Agent/SiteDirector.py#L1042

Commenting that line in code helped!

**Reply by aldbr** [reply]
Indeed, the `pythonVersion` option was deleted from `v8.0.41` from the `SiteDirector`, creating issues with all the previous versions.
So there are 2 solutions:
- commenting that line in Dirac `< v8.0.41` as you did.
- updating Dirac to a version `>= v8.0.41`.
